acasclient/selfile.py

```diff
@@ -0,0 +1,851 @@
+        :type file_str: str
+        :param file_type: File extension that indicates the file format.
+        :type file_type: Module constant
+        """
+
+        self._file_type = file_type
+        if type(file_str) == str:
+            self.file_name = io.StringIO(file_str)
+        else:
+            self.file_name = io.BytesIO(file_str)
+        read_func = get_read_func_by_file_type(self._file_type)
+        self._parse(read_func)
+
+    def saveAs(self, file_name):
+        """Save to a file on disk"""
+        raise NotImplementedError("Defined by subclass")
+
+    def validate(self):
+        raise NotImplementedError("Defined by subclass")
+
+    def getCorporateBatchIds(self):
+        """Return a set of Lot Corp Names recorded in the experiment"""
+
+        lot_corp_names = set()
+        if LOT_CORP_NAME in self.expt_df.columns:
+            lot_corp_names.update(self.expt_df[LOT_CORP_NAME].dropna().tolist())
+        return lot_corp_names
+
+    def getEndpointNames(self):
+        """Return set of Assay-Endpoints recorded in the experiment"""
+
+        endpoint_names = set(self.expt_df.columns.tolist())
+        endpoint_names.discard(LOT_CORP_NAME)
+        return endpoint_names
+
+
+    # Meta data property() decorators to flag when the representation
+    # is altered.
+    @property
+    def meta_has_changed(self):
+        """Return boolean,  True if the meta data changed after file parsing."""
+        return self._meta_has_changed
+    @meta_has_changed.setter
+    def meta_has_changed(self, value):
+        # Only set this if True so subsequent checks that are not changes
+        # can't unset.
+        if value:
+            self._meta_has_changed = value
+
+
+    @property
+    def experimental_meta_data(self):
+        """Experimental Meta Data (Meta Data)"""
+        return self._experimental_meta_data
+    @experimental_meta_data.setter
+    def experimental_meta_data(self, value):
+        self.meta_has_changed = (value != self._experimental_meta_data)
+        self._experimental_meta_data = value
+
+
+    @property
+    def format(self):
+        """Format (Meta Data)"""
+        return self._format
+    @format.setter
+    def format(self, value):
+        self.meta_has_changed = (value != self._format)
+        self._format = value
+
+
+    @property
+    def protocol_name(self):
+        """Protocol Name (Meta Data)"""
+        return self._protocol_name
+    @protocol_name.setter
+    def protocol_name(self, value):
+        self.meta_has_changed = (value != self._protocol_name)
+        self._protocol_name = value
+
+
+    @property
+    def experiment_name(self):
+        """Experiment Name (Meta Data)"""
+        return self._experiment_name
+    @experiment_name.setter
+    def experiment_name(self, value):
+        self.meta_has_changed = (value != self._experiment_name)
+        self._experiment_name = value
+    # Alias
+    name = experiment_name
```

Comment:
I added this alias for experiment name as we have a use case where all classes used in an ETL script are required to include a name property

Reply:
@bffrost should 'name' a formal property (with a setter/getter)?

Reply:
I guess I should test the setter but adding this as a setter did work.

Reply:
Alex and I actually discussed __slots__ vs decorators so the end user can not add more attributes.  I think adding more properties that don't map to a feature in an ACAS SEL file could lead to confusion.

Reply:
@brianbolt not to overcomplicate it, but you could implement the "name" field by wrapping this Experiment class in the ETL codebase. That'd keep acasclient "clean" since it is a little funny to have redundant names for the same attribute.

Reply:
All makes sense. I removed it here.